// @bazel/typescript/internal/tsetse/rules/conformance_pattern_rule.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../error_code", "../util/pattern_config", "../util/pattern_engines/property_write_engine"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    const error_code_1 = require("../error_code");
    const pattern_config_1 = require("../util/pattern_config");
    exports.PatternKind = pattern_config_1.PatternKind;
    const property_write_engine_1 = require("../util/pattern_engines/property_write_engine");
    /**
     * Builds a Rule that matches a certain pattern, given as parameter, and
     * that can additionally run a suggested fix generator on the matches.
     *
     * This is templated, mostly to ensure the nodes that have been matched
     * correspond to what the Fixer expects.
     */
    class ConformancePatternRule {
        constructor(config, fixer, verbose) {
            this.code = error_code_1.ErrorCode.CONFORMANCE_PATTERN;
            // TODO(rjamet): This cheats a bit with the typing, as TS doesn't realize
            // that P is Config.kind.
            // tslint:disable-next-line:no-any See above.
            let engine;
            switch (config.kind) {
                case pattern_config_1.PatternKind.BANNED_PROPERTY_WRITE:
                    engine = new property_write_engine_1.PropertyWriteEngine(config, fixer, verbose);
                    break;
                default:
                    throw new Error('Config type not recognized, or not implemented yet.');
            }
            this.ruleName = `conformance-pattern-${config.kind}`;
            this.engine = engine;
        }
        register(checker) {
            this.engine.register(checker);
        }
    }
    exports.ConformancePatternRule = ConformancePatternRule;
});
