// @bazel/typescript/internal/tsetse/failure.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    const ts = require("typescript");
    /**
     * A Tsetse check Failure is almost identical to a Diagnostic from TypeScript
     * except that:
     * (1) The error code is defined by each individual Tsetse rule.
     * (2) The optional `source` property is set to `Tsetse` so the host (VS Code
     * for instance) would use that to indicate where the error comes from.
     * (3) There's an optional suggestedFix field.
     */
    class Failure {
        constructor(sourceFile, start, end, failureText, code, suggestedFix) {
            this.sourceFile = sourceFile;
            this.start = start;
            this.end = end;
            this.failureText = failureText;
            this.code = code;
            this.suggestedFix = suggestedFix;
        }
        /**
         * This returns a structure compatible with ts.Diagnostic, but with added
         * fields, for convenience and to support suggested fixes.
         */
        toDiagnostic() {
            return {
                file: this.sourceFile,
                start: this.start,
                end: this.end,
                // start-end-using systems.
                length: this.end - this.start,
                messageText: this.failureText,
                category: ts.DiagnosticCategory.Error,
                code: this.code,
                // source is the name of the plugin.
                source: 'Tsetse',
                fix: this.suggestedFix
            };
        }
    }
    exports.Failure = Failure;
});
