// @applitools/dom-snapshot/src/browser/processResource.js
'use strict';
const filterInlineUrl = require('./filterInlineUrl');
const toUnAnchoredUri = require('./toUnAnchoredUri');
const absolutizeUrl = require('./absolutizeUrl');
const makeExtractResourcesFromStyle = require('./extractResourcesFromStyle');

function makeProcessResource({
  fetchUrl,
  findStyleSheetByUrl,
  extractResourcesFromStyleSheet,
  extractResourcesFromSvg,
  cache = {},
}) {
  let isFromSvgResource;
  const extractResourcesFromStyle = makeExtractResourcesFromStyle({extractResourcesFromStyleSheet});
  return function processResource(absoluteUrl, documents, baseUrl, getResourceUrlsAndBlobs) {
    return cache[absoluteUrl] || (cache[absoluteUrl] = doProcessResource(absoluteUrl));

    function doProcessResource(url) {
      return fetchUrl(url)
        .catch(e => {
          if (probablyCORS(e)) {
            return {probablyCORS: true, url};
          } else {
            throw e;
          }
        })
        .then(({url, type, value, probablyCORS}) => {
          if (probablyCORS) {
            return {resourceUrls: [url]};
          }

          let result = {blobsObj: {[url]: {type, value}}};
          let resourceUrls;
          if (/text\/css/.test(type)) {
            let styleSheet = findStyleSheetByUrl(url, documents);
            if (styleSheet || isFromSvgResource) {
              resourceUrls = extractResourcesFromStyle(value, styleSheet, documents[0]);
            }
          } else if (/image\/svg/.test(type)) {
            try {
              resourceUrls = extractResourcesFromSvg(value);
              if (resourceUrls && !isFromSvgResource) {
                isFromSvgResource = url;
              }
            } catch (e) {
              console.log('could not parse svg content', e);
            }
          }

          if (resourceUrls) {
            result = mapUrlsAndGetResult({resourceUrls, url, type, value}).then(
              res => (isFromSvgResource === url ? false : isFromSvgResource, res),
            );
          }
          return result;
        })
        .catch(err => {
          console.log('[dom-snapshot] error while fetching', url, err);
          return {};
        });
    }

    function mapUrlsAndGetResult({resourceUrls, url, type, value}) {
      const urls = resourceUrls
        .map(toUnAnchoredUri)
        .map(resourceUrl => absolutizeUrl(resourceUrl, url.replace(/^blob:/, '')))
        .filter(filterInlineUrl);
      return getResourceUrlsAndBlobs(documents, baseUrl, urls).then(({resourceUrls, blobsObj}) => ({
        resourceUrls,
        blobsObj: Object.assign(blobsObj, {[url]: {type, value}}),
      }));
    }

    function probablyCORS(err) {
      const msg =
        err.message &&
        (err.message.includes('Failed to fetch') || err.message.includes('Network request failed'));
      const name = err.name && err.name.includes('TypeError');
      return msg && name;
    }
  };
}

module.exports = makeProcessResource;
